Add REST API using grpc gateway

mayankjuneja opened the pull request:
Hi @YujiOshima, @jlewi,

We have been using Katib internally and one of the changes we have done is to expose core service as REST APIs (using [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway)). I think this might be useful for others as well (issues https://github.com/kubeflow/katib/issues/117, https://github.com/kubeflow/katib/issues/118). 

Side note - You might notice a lot of changes to vendor dir (these are mostly deletions). Doing a `dep ensure` cleaned up some of the unused packages. If there are other best practices around this, please let me know. I can also create a separate PR to clean up the unused packages.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/katib/142)
<!-- Reviewable:end -->

vinaykakade commented:
@YujiOshima - is there any advantage of checking-in the vendor folder with Katib? Checking in gopkg.lock and gopkg.toml should be sufficient. The PRs will be easier to look at that way, and any needed vendor packages can be downloaded locally based on lock/toml and without needing the vendor folder checked-in.

YujiOshima commented:
@mayankjuneja Thank you for the great contribution! I will take a look. Could you add some documents for REST?

@vinaykakade There is no exactry answer. https://github.com/golang/dep/blob/master/docs/FAQ.md#should-i-commit-my-vendor-directory
We can protect it from deletion of depending repository by including vendor folders. 
the more Katib mature, the less vendored folders change.

YujiOshima commented:
/ok-to-test

YujiOshima commented:
/assign YujiOshima

mayankjuneja commented on pkg/api/build.sh:
```diff
@@ -1 +1,3 @@
 docker run -it --rm -v $PWD:$(pwd) -w $(pwd) znly/protoc --python_out=plugins=grpc:./python --go_out=plugins=grpc:. -I. api.proto
+docker run -it --rm -v $PWD:$(pwd) -w $(pwd) znly/protoc --grpc-gateway_out=logtostderr=true:. -I. api.proto
```
building the grpc gateway and swagger stub files

mayankjuneja commented on scripts/build.sh:
```diff
@@ -28,6 +28,9 @@ cd ${SCRIPT_ROOT}
 docker build -t ${PREFIX}/vizier-core -f ${CMD_PREFIX}/manager/Dockerfile .
 
+echo "Building REST API for core image..."
+docker build -t ${PREFIX}/vizier-core-rest -f ${CMD_PREFIX}/manager_rest/Dockerfile .
```
build the image for manager REST API. it will be a separate service running in kubernetes.

mayankjuneja commented:
@YujiOshima I added some comments in the code and in this PR as well. Let me know if I should add more details.

YujiOshima commented on pkg/api/api.proto:
```diff
@@ -2,37 +2,148 @@ syntax = "proto3";
 }
 
 service Suggestion {
-    rpc GetSuggestions(GetSuggestionsRequest) returns (GetSuggestionsReply);
```
Do we need rest for suggestion and earlystopping service?

mayankjuneja replied:
thanks for pointing it out! it is not required.

YujiOshima commented:
@mayankjuneja  
* Please add build test for core-rest in CI.
https://github.com/kubeflow/katib/blob/ecbc5ddd47ce3979bdc39479eae5537a8593b6ef/test/scripts/run-tests.sh#L45

* Could you add doc to `doc` dir for how to use rest API with curl or other client?

Thank you!

mayankjuneja commented:
hi, wanted to give an update on this. I encountered an issue in our setup where the manager APIs become unresponsive when I start a study with a large number of trials (e.g. gridsearch with ~80 sets of parameter combinations). I am still investigating if this is related to grpc-gateway or underlying Katib APIs itself. I will get back to working on this PR once I have more details on it. Sorry for the delay on getting this out.

@jlewi @YujiOshima have you guys noticed any unusual behaviour with jobs with a higher number of trials?

jlewi commented:
@mayankjuneja We aren't running large hyperparameter sweeps; so I can't offer any comparison.

mayankjuneja commented:
@YujiOshima I did another pass and made the following changes:
1. added more documentation (in docs folder)
2. merged changes from master
3. updated the build, deploy scripts.

Can you review these changes? Thanks!

YujiOshima commented on manifests/vizier/core-rest/rbac.yaml:
```diff
@@ -0,0 +1,37 @@
+---
```
Since core-rest does not need to manage k8s resources, the rbac looks unnecessary.

mayankjuneja replied:
good point, thanks for catching this!

YujiOshima commented on docs/rest-api.md:
```diff
@@ -0,0 +1,77 @@
+For each RPC service, we define an equivalent HTTP REST API method using [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway). The mapping between service and REST API method can be found in this [file](https://github.com/kubeflow/katib/blob/master/pkg/api/api.proto). The mapping includes the URL path, query parameters and request body. You can read more details on the supported methods ...
+
+## Examples
+Assuming `{HOST}` as the ingress host for vizier-core-rest, 
```
Please add a way of using portforward.

kubectl v1.10~
`$ kubectl -n katib port-forward svc/vizier-core-rest 8080:80`

kubectl ~v1.9
`& kubectl -n katib port-forward $(kubectl -n katib get pod -o=name | grep vizier-core-rest | sed -e "s@pods\/@@") 8080:80 `

mayankjuneja replied:
I will add this.

YujiOshima commented on docs/rest-api.md:
````diff
@@ -0,0 +1,77 @@
+Request:
+
+```shell
+curl -X POST {HOST}/api/Manager/CreateStudy
````
Write explicitly to use above json file.
`curl -X POST {HOST}/api/Manager/CreateStudy -d @study_config.json`

mayankjuneja replied:
sorry, missed the argument here.

YujiOshima commented:
@mayankjuneja Thank for the great contribution!
Generally looks good. I checked it could work!
I commented a few points.
We can merge after fixing them.

mayankjuneja commented:
@YujiOshima thanks for the review! I added some changes based on your comments. Let me know if it looks good.

YujiOshima commented:
@mayankjuneja Thanks!
/lgtm
/approve

mayankjuneja commented:
/retest

YujiOshima commented:
@mayankjuneja I think you need to remove `serviceAccountName: vizier-core-rest` from core-rest deployment.

mayankjuneja commented:
@YujiOshima The tests passed now, can you please reapprove it?

YujiOshima commented:
🎉🎉🎉🎉🎉🎉🎉
